File: xaiop-sdk/python/src/xaiop/modes.py
"""Response consumption modes (multi-select). Default: callback only."""

from __future__ import annotations

from typing import Iterable

StreamMode = str

STREAM_MODES = {
    "CALLBACK": "callback",
    "PROMISE": "promise",
    "ASYNC_ITERATOR": "asyncIterator",
    "EVENTS": "events",
}

ALL_STREAM_MODES: frozenset[str] = frozenset(STREAM_MODES.values())


def normalize_modes(
    modes: Iterable[str] | str | None = None,
) -> set[str]:
    if modes is None:
        return {STREAM_MODES["CALLBACK"]}
    items = [modes] if isinstance(modes, str) else list(modes)
    out: set[str] = set()
    for m in items:
        if m not in ALL_STREAM_MODES:
            raise TypeError(f"unknown stream mode: {m!r}")
        out.add(m)
    if not out:
        out.add(STREAM_MODES["CALLBACK"])
    return out

File: xaiop-sdk/go/xaiop/modes.go
package xaiop

import "fmt"

// Stream mode constants (multi-select; default callback-only).
const (
	StreamModeCallback      = "callback"
	StreamModePromise       = "promise"
	StreamModeAsyncIterator = "asyncIterator"
	StreamModeEvents        = "events"
)

// NormalizeModes normalizes stream consumption modes.
func NormalizeModes(modes []string) (map[string]struct{}, error) {
	out := make(map[string]struct{})
	if len(modes) == 0 {
		out[StreamModeCallback] = struct{}{}
		return out, nil
	}
	for _, m := range modes {
		switch m {
		case StreamModeCallback, StreamModePromise, StreamModeAsyncIterator, StreamModeEvents:
			out[m] = struct{}{}
		default:
			return nil, fmt.Errorf("unknown stream mode: %q", m)
		}
	}
	if len(out) == 0 {
		out[StreamModeCallback] = struct{}{}
	}
	return out, nil
}
